src/modules/flight_mode_manager/tasks/AutoFollowTarget/FlightTaskAutoFollowTarget.cpp

```diff
@@ -64,266 +67,361 @@ bool FlightTaskAutoFollowTarget::activate(const vehicle_local_position_setpoint_
-	// altitude above the target. Clarify best solution.
-	switch (_param_nav_ft_alt_m.get()) {
-	case FOLLOW_ALTITUDE_MODE_TRACK_TARGET:
-		drone_desired_position(2) = target_position(2) - _param_nav_min_ft_ht.get();
-		break;
+	switch (_param_flw_tgt_alt_m.get()) {
+		case FOLLOW_ALTITUDE_MODE_TRACK_TERRAIN:
+			drone_desired_position(2) = ground_z_estimate - _follow_height;
+			break;
 
-	case FOLLOW_ALTITUDE_MODE_CONSTANT:
+		case FOLLOW_ALTITUDE_MODE_TRACK_TARGET:
+			drone_desired_position(2) = target_position(2) - _follow_height;
+			break;
 
-	// FALLTHROUGH
+		case FOLLOW_ALTITUDE_MODE_CONSTANT:
 
-	default:
-		// use the current position setpoint, unless it's closer to the ground than the minimum
-		// altitude setting
-		drone_desired_position(2) = math::min(_position_setpoint(2), -_param_nav_min_ft_ht.get());
+		// FALLTHROUGH
+
+		default:
+			// Calculate the desired Z position relative to the home position
+			drone_desired_position(2) = _home_position_z - _follow_height;
 	}
 
 	return drone_desired_position;
 }
 
-Vector3f FlightTaskAutoFollowTarget::calculate_target_position_filtered(Vector3f pos_ned_est, Vector3f vel_ned_est,
-		Vector3f acc_ned_est)
+
+float FlightTaskAutoFollowTarget::calculate_gimbal_height(const FollowAltitudeMode altitude_mode, const float target_pos_z) const
 {
-	// Reset the smoothness filter once the target position estimate is available
-	if (!PX4_ISFINITE(_target_position_filter.getState()(0)) || !PX4_ISFINITE(_target_position_filter.getState()(1))
-	    || !PX4_ISFINITE(_target_position_filter.getState()(2))) {
-		_target_position_filter.reset(pos_ned_est);
-	}
+	float gimbal_height{0.0f};
 
-	// Low-pass filter on target position.
-	_target_position_filter.setParameters(_deltatime, POSITION_FILTER_ALPHA);
+	switch (altitude_mode) {
+		case FOLLOW_ALTITUDE_MODE_TRACK_TERRAIN:
+			// Point the gimbal at the ground level in this tracking mode
+			gimbal_height = _dist_to_ground;
+			break;
 
-	if (_param_nav_ft_delc.get() == 0) {
-		_target_position_filter.update(pos_ned_est);
+		case FOLLOW_ALTITUDE_MODE_TRACK_TARGET:
+			// Point the gimbal at the target's 3D coordinates
+			gimbal_height = -(_position(2) - target_pos_z);
+			break;
 
-	} else {
-		// Use a predicted target's position to compensate the filter delay to some extent.
-		const Vector3f target_predicted_position = predict_future_pos_ned_est(POSITION_FILTER_ALPHA, pos_ned_est, vel_ned_est,
-				acc_ned_est);
-		_target_position_filter.update(target_predicted_position);
-	}
+		case FOLLOW_ALTITUDE_MODE_CONSTANT:
+
+		//FALLTHROUGH
 
-	return _target_position_filter.getState();
+		default:
+			gimbal_height = _home_position_z - _position(2); // Assume target is at home position's altitude
+			break;
+	}
 
+	return gimbal_height;
 }
 
+
 bool FlightTaskAutoFollowTarget::update()
 {
+	follow_target_status_s follow_target_status{}; // Debugging uORB message for follow target
+	// Members of the uORB message that gets set in the code below
```

Comment:
Agreed!

Reply:
Maybe it's just that I recently wrote a blog post on uORB so I am trying to be as user friendly as possible (in case someone doesn't know that follow_target_s struct implies that it's an uORB message)